File: VB/CircledProgressBarPainter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.Utils.Drawing
Imports System.Drawing

Namespace CircledProgressBarEditor
	Friend Class CircledProgressBarPainter
		Inherits ProgressBarPainter
		Protected Friend Shared Function CalcProportionalBorderRectangle(ByVal initRect As Rectangle, ByVal Proportion As Integer) As Rectangle
			Dim returnedRectangle As Rectangle = Rectangle.Empty
			initRect.Height -= 1
			If (initRect.Width \ initRect.Height) > Proportion Then
				Dim newWidth As Integer = initRect.Height * Proportion
				Dim newX As Integer = initRect.X + initRect.Width \ 2 - newWidth \ 2
				returnedRectangle = New Rectangle(newX, initRect.Y, newWidth, initRect.Height)
			Else
				returnedRectangle = initRect
			End If

			Return returnedRectangle
		End Function


		Protected Sub DrawEllipseAroundTheText(ByVal info As ControlGraphicsInfoArgs)
			Dim vi As CircledProgressBarViewInfo = TryCast(info.ViewInfo, CircledProgressBarViewInfo)
			Dim item As RepositoryItemCircledProgressBar = CType(vi.Item, RepositoryItemCircledProgressBar)
			If item.CustomBehavior.ShowTextEllipse Then
				Dim currentTextSize As SizeF = info.Graphics.MeasureString(vi.DisplayText, vi.Appearance.Font)
				If (vi.ContentRect.Height - currentTextSize.Height) > 20 Then
					Dim iRectForCaptionHeight As Integer = CInt(Fix(currentTextSize.Height)) + 15
					Dim iRectForCaptionWidth As Integer = vi.ContentRect.Width / (vi.ContentRect.Height \ iRectForCaptionHeight)
					Dim iRectForCaptionX As Integer = vi.ContentRect.X + vi.ContentRect.Width \ 2 - iRectForCaptionHeight \ 2
					Dim iRectForCaptionY As Integer = vi.ContentRect.Y + vi.ContentRect.Height \ 2 - iRectForCaptionHeight \ 2
					info.Graphics.FillEllipse(New SolidBrush(item.CustomBehavior.TextEllipseColor), New Rectangle(iRectForCaptionX, iRectForCaptionY, iRectForCaptionHeight, iRectForCaptionHeight))
				End If
			End If
		End Sub

		Protected Sub DrawCircledProgressBarString(ByVal info As ControlGraphicsInfoArgs)
			Dim vi As ProgressBarBaseViewInfo = TryCast(info.ViewInfo, ProgressBarBaseViewInfo)
			Dim emptyBrush As Brush = vi.ProgressAppearance.GetForeBrush(info.Cache)
			Dim fillBrush As Brush = vi.ProgressAppearance.GetFore2Brush(info.Cache)
			Dim clip As GraphicsClipState = info.Cache.ClipInfo.SaveAndSetClip(info.Bounds)
			Try
				If vi.ProgressInfo.ProgressBounds.Width = 0 Then
					info.Cache.DrawString(vi.DisplayText, vi.PaintAppearance.Font, emptyBrush, vi.ContentRect, vi.StrFormat)
					Return
				End If
				info.Cache.ClipInfo.ExcludeClip(vi.ProgressInfo.ProgressBounds)
				DrawEllipseAroundTheText(info)
				info.Cache.DrawString(vi.DisplayText, vi.PaintAppearance.Font, emptyBrush, vi.ContentRect, vi.StrFormat)
				info.Cache.ClipInfo.SetClip(vi.ProgressInfo.ProgressBounds)
				DrawEllipseAroundTheText(info)
				info.Cache.DrawString(vi.DisplayText, vi.PaintAppearance.Font, emptyBrush, vi.ContentRect, vi.StrFormat)

			Finally
				info.Cache.ClipInfo.RestoreClipRelease(clip)
			End Try
		End Sub

		Protected Overrides Sub DrawContent(ByVal info As ControlGraphicsInfoArgs)
			Dim vi As ProgressBarBaseViewInfo = TryCast(info.ViewInfo, ProgressBarBaseViewInfo)
			Dim item As RepositoryItemCircledProgressBar = CType(vi.Item, RepositoryItemCircledProgressBar)
			ObjectPainter.DrawObject(info.Cache, vi.ProgressPainter, vi.ProgressInfo)
			info.Graphics.DrawEllipse(New Pen(item.Appearance.BackColor), CalcProportionalBorderRectangle(info.Bounds, item.CustomBehavior.MinWidthHeightProportion))
			DrawCircledProgressBarString(info)
		End Sub

		Protected Overrides Sub DrawBorder(ByVal info As ControlGraphicsInfoArgs)
			Dim vi As ProgressBarBaseViewInfo = TryCast(info.ViewInfo, ProgressBarBaseViewInfo)
			Dim item As RepositoryItemCircledProgressBar = CType(vi.Item, RepositoryItemCircledProgressBar)
			If item.OwnerEdit IsNot Nothing AndAlso item.OwnerEdit.InplaceType = DevExpress.XtraEditors.Controls.InplaceType.Standalone Then
				info.Graphics.FillRectangle(New SolidBrush(item.OwnerEdit.Parent.BackColor), info.Bounds)
			End If
		End Sub

		Protected Overrides Function IsDrawBorderLast(ByVal info As ControlGraphicsInfoArgs) As Boolean
			Dim vi As ProgressBarBaseViewInfo = TryCast(info.ViewInfo, ProgressBarBaseViewInfo)
			If TypeOf vi.BorderPainter Is WindowsXPProgressBarBorderPainter Then
				Return False
			End If
			If TypeOf vi.ProgressPainter Is CircledProgressBarObjectPainter Then
				Return False
			End If
			Return True
		End Function
	End Class
End Namespace

File: CS/CircledProgressBarPainter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.Utils.Drawing;
using System.Drawing;

namespace CircledProgressBarEditor
{
    class CircledProgressBarPainter : ProgressBarPainter
    {
        protected internal static Rectangle CalcProportionalBorderRectangle(Rectangle initRect, int Proportion)
        {
            Rectangle returnedRectangle = Rectangle.Empty;
            initRect.Height--;
            if ((initRect.Width / initRect.Height) > Proportion)
            {
                int newWidth = initRect.Height * Proportion;
                int newX = initRect.X + initRect.Width / 2 - newWidth / 2;
                returnedRectangle = new Rectangle(newX, initRect.Y, newWidth, initRect.Height);
            }
            else
                returnedRectangle = initRect;

            return returnedRectangle;
        }


        protected void DrawEllipseAroundTheText(ControlGraphicsInfoArgs info)
        {
            CircledProgressBarViewInfo vi = info.ViewInfo as CircledProgressBarViewInfo;            
            RepositoryItemCircledProgressBar item = (RepositoryItemCircledProgressBar)vi.Item;
            if (item.CustomBehavior.ShowTextEllipse)
            {
                SizeF currentTextSize = info.Graphics.MeasureString(vi.DisplayText, vi.Appearance.Font);
                if ((vi.ContentRect.Height - currentTextSize.Height) > 20)
                {
                    int iRectForCaptionHeight = (int)currentTextSize.Height + 15;
                    int iRectForCaptionWidth = vi.ContentRect.Width / (vi.ContentRect.Height / iRectForCaptionHeight);
                    int iRectForCaptionX = vi.ContentRect.X + vi.ContentRect.Width / 2 - iRectForCaptionHeight / 2;
                    int iRectForCaptionY = vi.ContentRect.Y + vi.ContentRect.Height / 2 - iRectForCaptionHeight / 2;
                    info.Graphics.FillEllipse(new SolidBrush(item.CustomBehavior.TextEllipseColor),
                        new Rectangle(iRectForCaptionX, iRectForCaptionY, iRectForCaptionHeight, iRectForCaptionHeight));
                }
            }        
        }

        protected void DrawCircledProgressBarString(ControlGraphicsInfoArgs info)
        {
            ProgressBarBaseViewInfo vi = info.ViewInfo as ProgressBarBaseViewInfo;
            Brush emptyBrush = vi.ProgressAppearance.GetForeBrush(info.Cache);
            Brush fillBrush = vi.ProgressAppearance.GetFore2Brush(info.Cache);
            GraphicsClipState clip = info.Cache.ClipInfo.SaveAndSetClip(info.Bounds);
            try
            {
                if (vi.ProgressInfo.ProgressBounds.Width == 0)
                {
                    info.Cache.DrawString(vi.DisplayText, vi.PaintAppearance.Font, emptyBrush, vi.ContentRect, vi.StrFormat);
                    return;
                }
                info.Cache.ClipInfo.ExcludeClip(vi.ProgressInfo.ProgressBounds);
                DrawEllipseAroundTheText(info);
                info.Cache.DrawString(vi.DisplayText, vi.PaintAppearance.Font, emptyBrush, vi.ContentRect, vi.StrFormat);
                info.Cache.ClipInfo.SetClip(vi.ProgressInfo.ProgressBounds);
                DrawEllipseAroundTheText(info);
                info.Cache.DrawString(vi.DisplayText, vi.PaintAppearance.Font, emptyBrush, vi.ContentRect, vi.StrFormat);                

            }
            finally
            {
                info.Cache.ClipInfo.RestoreClipRelease(clip);
            }
        }

        protected override void DrawContent(ControlGraphicsInfoArgs info)
        {
            ProgressBarBaseViewInfo vi = info.ViewInfo as ProgressBarBaseViewInfo;
            RepositoryItemCircledProgressBar item = (RepositoryItemCircledProgressBar)vi.Item;            
            ObjectPainter.DrawObject(info.Cache, vi.ProgressPainter, vi.ProgressInfo);
            info.Graphics.DrawEllipse(new Pen(item.Appearance.BackColor), CalcProportionalBorderRectangle(info.Bounds, item.CustomBehavior.MinWidthHeightProportion));            
            DrawCircledProgressBarString(info);            
        }

        protected override void DrawBorder(ControlGraphicsInfoArgs info)
        {
            ProgressBarBaseViewInfo vi = info.ViewInfo as ProgressBarBaseViewInfo;
            RepositoryItemCircledProgressBar item = (RepositoryItemCircledProgressBar)vi.Item;
            if (item.OwnerEdit != null && item.OwnerEdit.InplaceType == DevExpress.XtraEditors.Controls.InplaceType.Standalone)
            {
                info.Graphics.FillRectangle(new SolidBrush(item.OwnerEdit.Parent.BackColor), info.Bounds);
            }
        }

        protected override bool IsDrawBorderLast(ControlGraphicsInfoArgs info)
        {
            ProgressBarBaseViewInfo vi = info.ViewInfo as ProgressBarBaseViewInfo;
            if (vi.BorderPainter is WindowsXPProgressBarBorderPainter) return false;
            if (vi.ProgressPainter is CircledProgressBarObjectPainter) return false;
            return true;
        }        
    }
}
